src/main/java/org/terasology/simpleFarming/systems/TreeAuthoritySystem.java

```diff
@@ -0,0 +1,573 @@
+        resetDelay(sapling, currentStage.minTime, currentStage.maxTime);
+	}
+	
+	/**
+	 * Grows the sapling into a tree using {@link #growSapling(SaplingDefinitionComponent)}.
+	 * 
+	 * @param event The delayed growth event.
+	 * @param sapling The sapling's block entity in the world, not used.
+	 * @param saplingComponent The sapling definition used to grow the tree.
+	 */
+	@ReceiveEvent
+    public void onSaplingGrowth(DelayedActionTriggeredEvent event, EntityRef sapling, SaplingDefinitionComponent saplingComponent) {
+		growSapling(saplingComponent);
+	}
+	
+	/**
+	 * Grows a tree based on a growth timer. This is done by first checking if the tree can be grown, then destroying the old one and growing a
+	 * new tree with the next growth stage. After that, the growth timer is reset if there is another stage.
+	 * 
+	 * @param event The delayed growth event.
+	 * @param rootEntity The block entity of the lowest log block in the tree, aka the "root".
+	 * @param logComponent The log component of the entity.
+	 * @param rootComponent The root component of the entity.
+	 * 
+	 * @see #canGenerateTree(EntityRef)
+	 * @see #destroyTree(EntityRef, Vector3i)
+	 * @see #generateTree(EntityRef)
+	 * @see #resetDelay(EntityRef, int, int)
+	 */
+	@ReceiveEvent
+	public void onRootGrowth(DelayedActionTriggeredEvent event, EntityRef rootEntity, LogComponent logComponent, RootComponent rootComponent) {
+		if(!rootComponent.alive || rootComponent.growthStage+1 == rootComponent.growthStages.size()) {
+			return;
+		}
+		
+		rootComponent.growthStage++;
+		if(canGenerateTree(rootEntity)) {
+			rootEntity.addOrSaveComponent(rootComponent);
+			destroyTree(rootEntity, logComponent.location);
+			generateTree(rootEntity);
+			if(rootComponent.growthStage+1 < rootComponent.growthStages.size()) {
+				TreeGrowthStage nextStage = rootComponent.growthStages.get(rootComponent.growthStage+1);
+				resetDelay(rootEntity, nextStage.minTime, nextStage.maxTime);
+			}
+		}
+	}
+	
+	/**
+	 * Reacts to an item with a CheatGrowthComponent being used on a sapling or log block, and appropriately grows or un-grows the tree.
+	 * 
+	 * @param event The event from right-clicking with an item.
+	 * @param item The item that was used.
+	 * @param cheatGrowthComponent The CheatGrowthComponent of that item.
+	 * @param itemComponent The ItemComponent of that item, not used.
+	 * 
+	 * @see #growSapling(SaplingDefinitionComponent)
+	 * @see #onRootGrowth(DelayedActionTriggeredEvent, EntityRef, LogComponent, RootComponent)
+	 */
+	@ReceiveEvent
+	public void onCheatGrowth(ActivateEvent event, EntityRef item, CheatGrowthComponent cheatGrowthComponent, ItemComponent itemComponent) {
+		EntityRef target = event.getTarget();
+		if(target.hasComponent(SaplingDefinitionComponent.class)) {
+			if(!cheatGrowthComponent.causesUnGrowth) {
+				growSapling(target.getComponent(SaplingDefinitionComponent.class));
+			}
+		} else if(target.hasComponent(LogComponent.class)) {
```

Comment:
Ideally this section of the if-statement should be split off into it's own method.

You could probably reuse the other methods in this system for the code in here. You've already split other methods up so why not re-use them here ;)

In particular my gut is telling me you might be able to tweak the `OnRootGrowth` and/or `generate/destroyTree` methods for both growing and un-growing (for ungrowing simply step back two stages and then grow forwards one?)

Reply:
I think I fixed this in the latest commit.